### DISCUSSION TITLE
Tesseract now available on Python PyPi (pip)!

### DISCUSSION THREAD

**Discussion by johnwason** [tesseract-robotics/tesseract#754] [General]
Tesseract Robotics is now available on PyPI, and can be installed using `pip`.

Currently only CPython 3.9 x64 on Windows (and now on Ubuntu 20.04 Python 3.8 too!) is available. I can try to add more Python versions and possibly Linux (done! see below) in the future.

To install on Windows:

Install Python 3.9 for all users (C:\Python39 install directory): https://www.python.org/ftp/python/3.9.12/python-3.9.12-amd64.exe

edit: now also available for Python 3.8 and Python 3.10 on Windows

Run the following command:

```
python -m pip install tesseract-robotics tesseract-robotics-viewer
```

To install on Ubuntu 20.04:

```
sudo apt install python3-pip python3-numpy
# The supplied version of pip is too old for manylinux_2_31, upgrade pip
python3 -m pip install -U pip
python3 -m pip install --user  tesseract_robotics tesseract_robotics_viewer
```

Running an example requires files from the GitHub repositories. These are not required to use the package They are only resource files like URDF, meshes, and the example script itself.

To run the example on Windows:

```
git clone --depth=1 https://github.com/tesseract-robotics/tesseract.git

git clone --depth=1 --branch pr/UpstreamChanges5 https://github.com/johnwason/tesseract_python-1.git tesseract_python

set TESSERACT_SUPPORT_DIR=%CD%\tesseract\tesseract_support

cd tesseract_python\tesseract_viewer_python\examples

python abb_irb2400_viewer.py

```

New (or updated) tools like `auditwheel` and the new `manylinux_x_y` platforms have made working with Linux wheels and dependencies much easier.

~~Portability of Python wheels can be tricky, so I am not sure if this will work on every computer. This build uses shared libraries, which isn't ideal since there can be version and symbol conflicts (aka DLL hell). Ideally static linking should be used for wheels, but it is significantly more difficult and will require significantly more time.~~

~~`vcpkg` was used to build the dependencies that aren't included in the ROS-I projects.~~

Let me know if you have any issues!

**Comment by Levi-Armstrong** [comment]
What would it take to get it released on linux?

**Reply by savolla** [reply]
Great news! Can't wait the Linux version!

**Reply by johnwason** [reply]
PyPI uses "manylinux" builds, which are pretty tricky. The current one `manylinux2014` only works on Ubuntu 20.04 or greater. https://github.com/pypa/manylinux

I'll give it a try this afternoon and see if it works.

**Reply by johnwason** [reply]
No luck, building a `manylinux` wheel is going to be very difficult. The dependencies for Tesseract use system libraries that can't be used inside a `manylinux` project. It may be eventually possible, but is probably not worth the effort. It would make more sense to set up a PPA or private Ubuntu repository. I have done both, and it isn't terribly difficult.

**Reply by johnwason** [reply]
Here is my private Debian repository for Robot Raconteur: https://github.com/robotraconteur/robotraconteur-apt

Instructions here: https://github.com/robotraconteur/robotraconteur#debian-10-buster

**Reply by savolla** [reply]
Will definitely check this out thanks!

**Reply by johnwason** [reply]
I was wrong about Linux wheels. There is a tool called `auditwheel` which can automatically gather the shared libraries and include them in the package. There is also a `manylinux_2_31` platform version which corresponds directly to Ubuntu 20.04. This means that the output from a normal build can be packaged directly with the following commands:

```
python3 setup.py bdist_wheel --plat-name=manylinux_2_31_x86_64
python3 -m auditwheel repair --plat manylinux_2_31_x86_64 dist/tesseract_robotics-0.1.0-cp38-cp38-manylinux_2_31_x86_64.whl
```

The resulting wheel can be uploaded to pypi.org. Currently the plugins for kinematics and collision are not included, since there is no loader link that `auditwheel` can follow. We'll have to add an option to compile-time link these plugins, but it shouldn't be too difficult.

I uploaded a test package to testpypi.org. This package doesn't have any plugins so you can't do much, but it does load and the non-plugin contents seem to work. Install using:

```
sudo apt install python3-pip python3-numpy
# The supplied version of pip is too old for manylinux_2_31, upgrade pip
python3 -m pip install -U pip
python3 -m pip install --user --index-url https://test.pypi.org/simple/ tesseract_robotics
```

**Reply by johnwason** [reply]
The Python wheel for Ubuntu 20.04 is ready! See the edited original post for details.
